File: tests/scala/src/test/scala/sertests/TestsBase.scala
package sertests

import java.io.IOException
import zio.test._
import zio.test.Assertion._
import zio.IO

abstract class TestsBase extends DefaultRunnableSpec {

    protected def testCases: Seq[TestCase[_]]

    private def encodeTest[A](testCase: TestCase[A]): IO[IOException, TestResult] =
        for {
            writer <- MemoryFormatWriter.make
            _ <- testCase.codec.write(writer, testCase.value)
            actual <- writer.toChunk
        } yield assert(actual)(equalTo(testCase.encoded))

    private def decodeTest[A](testCase: TestCase[A]): IO[IOException, TestResult] =
        for {
            reader <- MemoryFormatReader.fromChunk(testCase.encoded)
            actual <- testCase.codec.read(reader)
            eof <- reader.isEOF
        } yield assert(eof)(isTrue) && assert(actual)(equalTo(testCase.value))


    override def spec: ZSpec[Environment, Failure] =
        suite("Verilization tests")(testCases.zipWithIndex.flatMap { case (testCase, i) =>
            Seq(
                testM(s"test $i encode")(encodeTest(testCase)),
                testM(s"test $i decode")(decodeTest(testCase)),
            )
        }: _*)
}

File: tests/java/src/test/java/sertests/TestsBase.java
package sertests;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.Arrays;

import dev.argon.verilization.runtime.Codec;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class TestsBase {
    protected <T> void check(Codec<T> codec, T value, byte[] encoded) throws IOException {
        {
            final var writer = new MemoryFormatWriter();
            codec.write(writer, value);
            assertArrayEquals(writer.toByteArray(), encoded);
        }

        {
            final var reader = new MemoryFormatReader(ByteBuffer.wrap(encoded));
            final var decoded = codec.read(reader);
            assertTrue(reader.isEOF());
            assertEquals(decoded, value);
            assertEquals(value, decoded);
        }
    }
}
